File: src/domain/onboarding/repo.ts
import type { PoolClient } from 'pg';
import { query } from '../../db/pool';
import type {
  ApprovedAccountSummary,
  BusinessDetails,
  DirectorOrBeneficialOwner,
  OnboardingApplication,
  OnboardingStatus,
  RejectionDetail,
  UploadedDocument,
} from '../../contract';
import type { PrincipalInput } from './schemas';

interface ApplicationRow {
  id: string;
  customer_id: string;
  status: OnboardingStatus;
  business: BusinessDetails | null;
  rejection_reasons: RejectionDetail[] | null;
  approved_summary: ApprovedAccountSummary | null;
  submitted_at: Date | null;
  reviewed_at: Date | null;
}

interface PrincipalRow {
  id: string;
  full_name: string;
  role: DirectorOrBeneficialOwner['role'];
  ownership_percentage: string | null;
  date_of_birth: string | null;
  bvn: string | null;
  nin: string | null;
}

interface DocumentRow {
  id: string;
  type: UploadedDocument['type'];
  file_name: string;
  mime_type: string;
  size_bytes: string;
  status: UploadedDocument['status'];
  upload_progress_percent: number;
  uploaded_at: Date | null;
  error_message: string | null;
}

function iso(value: Date | null): string | undefined {
  return value ? value.toISOString() : undefined;
}

function toPrincipal(row: PrincipalRow): DirectorOrBeneficialOwner {
  return {
    id: row.id,
    fullName: row.full_name,
    role: row.role,
    ...(row.ownership_percentage !== null
      ? { ownershipPercentage: Number(row.ownership_percentage) }
      : {}),
    ...(row.date_of_birth !== null ? { dateOfBirth: row.date_of_birth } : {}),
    ...(row.bvn !== null ? { bvn: row.bvn } : {}),
    ...(row.nin !== null ? { nin: row.nin } : {}),
  };
}

function toDocument(row: DocumentRow): UploadedDocument {
  return {
    id: row.id,
    type: row.type,
    fileName: row.file_name,
    mimeType: row.mime_type,
    sizeBytes: Number(row.size_bytes),
    status: row.status,
    uploadProgressPercent: row.upload_progress_percent,
    ...(row.uploaded_at ? { uploadedAt: row.uploaded_at.toISOString() } : {}),
    ...(row.error_message !== null ? { errorMessage: row.error_message } : {}),
  };
}

// Any pg querier — the pool (module `query`) or a transaction client. Reads
// inside a transaction MUST use the client so they see uncommitted rows.
type Querier = <T>(text: string, params?: readonly unknown[]) => Promise<{ rows: T[] }>;

const poolQuerier: Querier = (text, params) =>
  query(text, params) as unknown as Promise<{ rows: never[] }>;

async function assembleFrom(
  appRow: ApplicationRow,
  q: Querier = poolQuerier,
): Promise<OnboardingApplication> {
  const [principals, documents] = await Promise.all([
    q<PrincipalRow>(
      `select id, full_name, role, ownership_percentage,
              date_of_birth::text as date_of_birth, bvn, nin
         from onboarding_principals
        where application_id = $1
        order by position`,
      [appRow.id],
    ),
    q<DocumentRow>(
      `select id, type, file_name, mime_type, size_bytes, status,
              upload_progress_percent, uploaded_at, error_message
         from onboarding_documents
        where application_id = $1
        order by created_at`,
      [appRow.id],
    ),
  ]);

  return {
    id: appRow.id,
    customerId: appRow.customer_id,
    status: appRow.status,
    business: appRow.business,
    principals: principals.rows.map(toPrincipal),
    documents: documents.rows.map(toDocument),
    ...(appRow.rejection_reasons ? { rejectionReasons: appRow.rejection_reasons } : {}),
    ...(appRow.approved_summary ? { approvedSummary: appRow.approved_summary } : {}),
    ...(iso(appRow.submitted_at) ? { submittedAt: iso(appRow.submitted_at) } : {}),
    ...(iso(appRow.reviewed_at) ? { reviewedAt: iso(appRow.reviewed_at) } : {}),
  };
}

const APP_COLUMNS = `id, customer_id, status, business, rejection_reasons,
                     approved_summary, submitted_at, reviewed_at`;

export async function findApplicationByCustomer(
  customerId: string,
): Promise<OnboardingApplication | null> {
  const { rows } = await query<ApplicationRow>(
    `select ${APP_COLUMNS} from onboarding_applications where customer_id = $1`,
    [customerId],
  );
  return rows[0] ? assembleFrom(rows[0]) : null;
}

export async function findApplicationById(id: string): Promise<OnboardingApplication | null> {
  const { rows } = await query<ApplicationRow>(
    `select ${APP_COLUMNS} from onboarding_applications where id = $1`,
    [id],
  );
  return rows[0] ? assembleFrom(rows[0]) : null;
}

/** For the transaction path — re-reads within the caller's client. */
export async function reloadApplication(
  client: PoolClient,
  id: string,
): Promise<OnboardingApplication> {
  const { rows } = await client.query<ApplicationRow>(
    `select ${APP_COLUMNS} from onboarding_applications where id = $1`,
    [id],
  );
  const row = rows[0];
  if (!row) throw new Error(`application ${id} vanished mid-transaction`);
  const clientQuerier: Querier = (text, params) =>
    client.query(text, params as unknown[]) as unknown as Promise<{ rows: never[] }>;
  return assembleFrom(row, clientQuerier);
}

export async function saveBusiness(
  client: PoolClient,
  applicationId: string,
  business: BusinessDetails,
): Promise<void> {
  await client.query(
    `update onboarding_applications
        set business = $2::jsonb,
            status = case when status = 'draft' then 'draft' else status end,
            updated_at = now()
      where id = $1`,
    [applicationId, JSON.stringify(business)],
  );
}

export async function replacePrincipals(
  client: PoolClient,
  applicationId: string,
  principals: readonly PrincipalInput[],
): Promise<void> {
  await client.query(`delete from onboarding_principals where application_id = $1`, [applicationId]);
  for (let i = 0; i < principals.length; i++) {
    const p = principals[i]!;
    await client.query(
      `insert into onboarding_principals
         (application_id, position, full_name, role, ownership_percentage, date_of_birth, bvn, nin)
       values ($1, $2, $3, $4, $5, $6, $7, $8)`,
      [
        applicationId,
        i,
        p.fullName,
        p.role,
        p.ownershipPercentage ?? null,
        p.dateOfBirth ?? null,
        p.bvn ?? null,
        p.nin ?? null,
      ],
    );
  }
  await client.query(
    `update onboarding_applications set updated_at = now() where id = $1`,
    [applicationId],
  );
}

export interface DocumentRecordInput {
  type: UploadedDocument['type'];
  fileName: string;
  mimeType: string;
  sizeBytes: number;
  storageKey: string;
}

export async function upsertUploadedDocument(
  client: PoolClient,
  applicationId: string,
  input: DocumentRecordInput,
): Promise<UploadedDocument> {
  const { rows } = await client.query<DocumentRow>(
    `insert into onboarding_documents
       (application_id, type, file_name, mime_type, size_bytes, status,
        upload_progress_percent, storage_key, uploaded_at)
     values ($1, $2, $3, $4, $5, 'uploaded', 100, $6, now())
     on conflict (application_id, type) do update
       set file_name = excluded.file_name,
           mime_type = excluded.mime_type,
           size_bytes = excluded.size_bytes,
           status = 'uploaded',
           upload_progress_percent = 100,
           storage_key = excluded.storage_key,
           uploaded_at = now(),
           error_message = null
     returning id, type, file_name, mime_type, size_bytes, status,
               upload_progress_percent, uploaded_at, error_message`,
    [applicationId, input.type, input.fileName, input.mimeType, input.sizeBytes, input.storageKey],
  );
  return toDocument(rows[0]!);
}

interface DocumentWithKeyRow extends DocumentRow {
  storage_key: string | null;
  application_id: string;
}

export async function findDocument(
  documentId: string,
): Promise<{ document: UploadedDocument; applicationId: string; storageKey: string | null } | null> {
  const { rows } = await query<DocumentWithKeyRow>(
    `select id, application_id, type, file_name, mime_type, size_bytes, status,
            upload_progress_percent, uploaded_at, error_message, storage_key
       from onboarding_documents
      where id = $1`,
    [documentId],
  );
  const row = rows[0];
  if (!row) return null;
  return { document: toDocument(row), applicationId: row.application_id, storageKey: row.storage_key };
}

export async function markDocumentUploaded(
  client: PoolClient,
  documentId: string,
): Promise<UploadedDocument> {
  const { rows } = await client.query<DocumentRow>(
    `update onboarding_documents
        set status = 'uploaded', upload_progress_percent = 100,
            error_message = null, uploaded_at = now()
      where id = $1
     returning id, type, file_name, mime_type, size_bytes, status,
               upload_progress_percent, uploaded_at, error_message`,
    [documentId],
  );
  return toDocument(rows[0]!);
}

export async function deleteDocument(client: PoolClient, documentId: string): Promise<void> {
  await client.query(`delete from onboarding_documents where id = $1`, [documentId]);
}

export interface StatusPatch {
  status: OnboardingStatus;
  submittedAt?: boolean;
  reviewedAt?: boolean;
  /** Always written: pass the summary to set it, omit to clear it. */
  approvedSummary?: ApprovedAccountSummary;
  /** Always written: pass reasons to set them, omit to clear them. */
  rejectionReasons?: readonly RejectionDetail[];
}

/**
 * Authoritative status write. `approved_summary` and `rejection_reasons` are
 * set to exactly what the patch carries (omitted => NULL), so a resubmit after
 * a rejection clears the stale reasons and vice versa. Only `submit` calls
 * this, and it always passes the complete intended picture.
 */
export async function patchStatus(
  client: PoolClient,
  applicationId: string,
  patch: StatusPatch,
): Promise<void> {
  await client.query(
    `update onboarding_applications
        set status = $2,
            submitted_at = case when $3::boolean then now() else submitted_at end,
            reviewed_at = case when $4::boolean then now() else reviewed_at end,
            approved_summary = $5::jsonb,
            rejection_reasons = $6::jsonb,
            updated_at = now()
      where id = $1`,
    [
      applicationId,
      patch.status,
      patch.submittedAt ?? false,
      patch.reviewedAt ?? false,
      patch.approvedSummary ? JSON.stringify(patch.approvedSummary) : null,
      patch.rejectionReasons && patch.rejectionReasons.length
        ? JSON.stringify(patch.rejectionReasons)
        : null,
    ],
  );
}

export async function replaceKybChecks(
  client: PoolClient,
  applicationId: string,
  checks: ReadonlyArray<{ key: string; passed: boolean; detail?: string }>,
): Promise<void> {
  await client.query(`delete from kyb_checks where application_id = $1`, [applicationId]);
  for (const check of checks) {
    await client.query(
      `insert into kyb_checks (application_id, check_key, passed, detail)
       values ($1, $2, $3, $4)`,
      [applicationId, check.key, check.passed, check.detail ?? null],
    );
  }
}

File: src/domain/onboarding/repo.rs
use crate::contract::onboarding::{
    ApprovedAccountSummary, BusinessDetails, DirectorOrBeneficialOwner, OnboardingApplication,
    PrincipalRole, RejectionDetail, UploadedDocument,
};
use crate::error::ApiResult;
use crate::util::{iso, iso_opt};
use chrono::{DateTime, Utc};
use sqlx::types::Json;
use sqlx::PgPool;
use uuid::Uuid;

#[derive(sqlx::FromRow)]
struct AppRow {
    id: Uuid,
    customer_id: Uuid,
    status: String,
    business: Option<Json<BusinessDetails>>,
    rejection_reasons: Option<Json<Vec<RejectionDetail>>>,
    approved_summary: Option<Json<ApprovedAccountSummary>>,
    submitted_at: Option<DateTime<Utc>>,
    reviewed_at: Option<DateTime<Utc>>,
}

#[derive(sqlx::FromRow)]
struct PrincipalRow {
    id: Uuid,
    full_name: String,
    role: String,
    ownership_percentage: Option<f64>,
    date_of_birth: Option<String>,
    bvn: Option<String>,
    nin: Option<String>,
}

#[derive(sqlx::FromRow)]
struct DocRow {
    id: Uuid,
    #[sqlx(rename = "type")]
    doc_type: String,
    file_name: String,
    mime_type: String,
    size_bytes: i64,
    status: String,
    upload_progress_percent: i32,
    uploaded_at: Option<DateTime<Utc>>,
    error_message: Option<String>,
}

impl DocRow {
    fn into_contract(self) -> UploadedDocument {
        UploadedDocument {
            id: self.id.to_string(),
            r#type: self.doc_type,
            file_name: self.file_name,
            mime_type: self.mime_type,
            size_bytes: self.size_bytes,
            status: self.status,
            upload_progress_percent: self.upload_progress_percent,
            uploaded_at: iso_opt(self.uploaded_at),
            error_message: self.error_message,
        }
    }
}

const APP_COLS: &str = "id, customer_id, status, business, rejection_reasons,
                        approved_summary, submitted_at, reviewed_at";

async fn assemble(pool: &PgPool, app: AppRow) -> ApiResult<OnboardingApplication> {
    let principals: Vec<PrincipalRow> = sqlx::query_as(
        "select id, full_name, role, ownership_percentage,
                date_of_birth::text as date_of_birth, bvn, nin
           from onboarding_principals where application_id = $1 order by position",
    )
    .bind(app.id)
    .fetch_all(pool)
    .await?;

    let documents: Vec<DocRow> = sqlx::query_as(
        "select id, type, file_name, mime_type, size_bytes, status,
                upload_progress_percent, uploaded_at, error_message
           from onboarding_documents where application_id = $1 order by created_at",
    )
    .bind(app.id)
    .fetch_all(pool)
    .await?;

    Ok(OnboardingApplication {
        id: app.id.to_string(),
        customer_id: app.customer_id.to_string(),
        status: app.status,
        business: app.business.map(|j| j.0),
        principals: principals
            .into_iter()
            .map(|p| DirectorOrBeneficialOwner {
                id: Some(p.id.to_string()),
                full_name: p.full_name,
                role: PrincipalRole::parse(&p.role).unwrap_or(PrincipalRole::Director),
                ownership_percentage: p.ownership_percentage,
                date_of_birth: p.date_of_birth,
                bvn: p.bvn,
                nin: p.nin,
            })
            .collect(),
        documents: documents.into_iter().map(DocRow::into_contract).collect(),
        rejection_reasons: app.rejection_reasons.map(|j| j.0),
        approved_summary: app.approved_summary.map(|j| j.0),
        submitted_at: app.submitted_at.map(iso),
        reviewed_at: app.reviewed_at.map(iso),
    })
}

pub async fn find_by_customer(
    pool: &PgPool,
    customer_id: Uuid,
) -> ApiResult<Option<OnboardingApplication>> {
    let row: Option<AppRow> = sqlx::query_as(&format!(
        "select {APP_COLS} from onboarding_applications where customer_id = $1"
    ))
    .bind(customer_id)
    .fetch_optional(pool)
    .await?;
    match row {
        Some(r) => Ok(Some(assemble(pool, r).await?)),
        None => Ok(None),
    }
}

pub async fn save_business(
    conn: &mut sqlx::PgConnection,
    app_id: Uuid,
    business: &BusinessDetails,
) -> ApiResult<()> {
    sqlx::query(
        "update onboarding_applications
            set business = $2, updated_at = now()
          where id = $1",
    )
    .bind(app_id)
    .bind(Json(business))
    .execute(conn)
    .await?;
    Ok(())
}

pub async fn replace_principals(
    conn: &mut sqlx::PgConnection,
    app_id: Uuid,
    principals: &[DirectorOrBeneficialOwner],
) -> ApiResult<()> {
    sqlx::query("delete from onboarding_principals where application_id = $1")
        .bind(app_id)
        .execute(&mut *conn)
        .await?;
    for (i, p) in principals.iter().enumerate() {
        sqlx::query(
            "insert into onboarding_principals
               (application_id, position, full_name, role, ownership_percentage, date_of_birth, bvn, nin)
             values ($1, $2, $3, $4, $5, $6::date, $7, $8)",
        )
        .bind(app_id)
        .bind(i as i32)
        .bind(&p.full_name)
        .bind(p.role.as_str())
        .bind(p.ownership_percentage)
        .bind(p.date_of_birth.as_deref())
        .bind(p.bvn.as_deref())
        .bind(p.nin.as_deref())
        .execute(&mut *conn)
        .await?;
    }
    sqlx::query("update onboarding_applications set updated_at = now() where id = $1")
        .bind(app_id)
        .execute(&mut *conn)
        .await?;
    Ok(())
}

pub struct DocInput<'a> {
    pub doc_type: &'a str,
    pub file_name: &'a str,
    pub mime_type: &'a str,
    pub size_bytes: i64,
    pub storage_key: &'a str,
}

pub async fn upsert_document(
    conn: &mut sqlx::PgConnection,
    app_id: Uuid,
    input: DocInput<'_>,
) -> ApiResult<UploadedDocument> {
    let row: DocRow = sqlx::query_as(
        "insert into onboarding_documents
           (application_id, type, file_name, mime_type, size_bytes, status,
            upload_progress_percent, storage_key, uploaded_at)
         values ($1, $2, $3, $4, $5, 'uploaded', 100, $6, now())
         on conflict (application_id, type) do update
           set file_name = excluded.file_name, mime_type = excluded.mime_type,
               size_bytes = excluded.size_bytes, status = 'uploaded',
               upload_progress_percent = 100, storage_key = excluded.storage_key,
               uploaded_at = now(), error_message = null
         returning id, type, file_name, mime_type, size_bytes, status,
                   upload_progress_percent, uploaded_at, error_message",
    )
    .bind(app_id)
    .bind(input.doc_type)
    .bind(input.file_name)
    .bind(input.mime_type)
    .bind(input.size_bytes)
    .bind(input.storage_key)
    .fetch_one(conn)
    .await?;
    Ok(row.into_contract())
}

pub struct FoundDocument {
    pub document: UploadedDocument,
    pub application_id: Uuid,
    pub storage_key: Option<String>,
}

#[derive(sqlx::FromRow)]
struct DocWithMetaRow {
    id: Uuid,
    application_id: Uuid,
    #[sqlx(rename = "type")]
    doc_type: String,
    file_name: String,
    mime_type: String,
    size_bytes: i64,
    status: String,
    upload_progress_percent: i32,
    uploaded_at: Option<DateTime<Utc>>,
    error_message: Option<String>,
    storage_key: Option<String>,
}

pub async fn find_document(pool: &PgPool, doc_id: Uuid) -> ApiResult<Option<FoundDocument>> {
    let row: Option<DocWithMetaRow> = sqlx::query_as(
        "select id, application_id, type, file_name, mime_type, size_bytes, status,
                upload_progress_percent, uploaded_at, error_message, storage_key
           from onboarding_documents where id = $1",
    )
    .bind(doc_id)
    .fetch_optional(pool)
    .await?;
    Ok(row.map(|r| FoundDocument {
        document: UploadedDocument {
            id: r.id.to_string(),
            r#type: r.doc_type,
            file_name: r.file_name,
            mime_type: r.mime_type,
            size_bytes: r.size_bytes,
            status: r.status,
            upload_progress_percent: r.upload_progress_percent,
            uploaded_at: iso_opt(r.uploaded_at),
            error_message: r.error_message,
        },
        application_id: r.application_id,
        storage_key: r.storage_key,
    }))
}

pub async fn mark_document_uploaded(
    conn: &mut sqlx::PgConnection,
    doc_id: Uuid,
) -> ApiResult<UploadedDocument> {
    let row: DocRow = sqlx::query_as(
        "update onboarding_documents
            set status = 'uploaded', upload_progress_percent = 100,
                error_message = null, uploaded_at = now()
          where id = $1
        returning id, type, file_name, mime_type, size_bytes, status,
                  upload_progress_percent, uploaded_at, error_message",
    )
    .bind(doc_id)
    .fetch_one(conn)
    .await?;
    Ok(row.into_contract())
}

pub async fn delete_document(conn: &mut sqlx::PgConnection, doc_id: Uuid) -> ApiResult<()> {
    sqlx::query("delete from onboarding_documents where id = $1")
        .bind(doc_id)
        .execute(conn)
        .await?;
    Ok(())
}

#[derive(Default)]
pub struct StatusPatch {
    pub status: String,
    pub mark_submitted: bool,
    pub mark_reviewed: bool,
    pub approved_summary: Option<ApprovedAccountSummary>,
    pub rejection_reasons: Option<Vec<RejectionDetail>>,
}

/// Authoritative status write — approved_summary / rejection_reasons are set to
/// exactly what the patch carries (None => NULL), so resubmit after a rejection
/// clears stale reasons.
pub async fn patch_status(
    conn: &mut sqlx::PgConnection,
    app_id: Uuid,
    patch: StatusPatch,
) -> ApiResult<()> {
    sqlx::query(
        "update onboarding_applications
            set status = $2,
                submitted_at = case when $3 then now() else submitted_at end,
                reviewed_at  = case when $4 then now() else reviewed_at end,
                approved_summary  = $5,
                rejection_reasons = $6,
                updated_at = now()
          where id = $1",
    )
    .bind(app_id)
    .bind(&patch.status)
    .bind(patch.mark_submitted)
    .bind(patch.mark_reviewed)
    .bind(patch.approved_summary.map(Json))
    .bind(patch.rejection_reasons.filter(|r| !r.is_empty()).map(Json))
    .execute(conn)
    .await?;
    Ok(())
}

pub async fn replace_kyb_checks(
    conn: &mut sqlx::PgConnection,
    app_id: Uuid,
    checks: &[super::kyb::CheckResult],
) -> ApiResult<()> {
    sqlx::query("delete from kyb_checks where application_id = $1")
        .bind(app_id)
        .execute(&mut *conn)
        .await?;
    for c in checks {
        sqlx::query(
            "insert into kyb_checks (application_id, check_key, passed, detail)
             values ($1, $2, $3, $4)",
        )
        .bind(app_id)
        .bind(c.key)
        .bind(c.passed)
        .bind(&c.detail)
        .execute(&mut *conn)
        .await?;
    }
    Ok(())
}
